[GTK][WPE] Implement KeyboardEvent.repeat

carlosgcampos opened the pull request:
#### 416736dee2ceb916e036585ff30443d49b65d90f
<pre>
[GTK][WPE] Implement KeyboardEvent.repeat
<a href="https://bugs.webkit.org/show_bug.cgi?id=253507">https://bugs.webkit.org/show_bug.cgi?id=253507</a>

Reviewed by Michael Catanzaro.

It&apos;s always false for GTK and WPE. This is also used since 259146@main
to decide when to use the keyboard scrolling animation for page scrolls.
Neither GTK nor libpwe provide information about whether a key press
event is auto-repeated or not, so this adds a simple helper class to
handle it.

* Source/WebKit/Shared/NativeWebKeyboardEvent.h:
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebKeyboardEvent.cpp:
(WebKit::WebKeyboardEvent::WebKeyboardEvent):
* Source/WebKit/Shared/WebKeyboardEvent.h:
* Source/WebKit/Shared/gtk/NativeWebKeyboardEventGtk.cpp:
(WebKit::NativeWebKeyboardEvent::NativeWebKeyboardEvent):
* Source/WebKit/Shared/gtk/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebKeyboardEvent):
* Source/WebKit/Shared/gtk/WebEventFactory.h:
* Source/WebKit/Shared/libwpe/NativeWebKeyboardEventLibWPE.cpp:
(WebKit::NativeWebKeyboardEvent::NativeWebKeyboardEvent):
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebKeyboardEvent):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/KeyAutoRepeatHandler.cpp: Added.
(WebKit::KeyAutoRepeatHandler::keyPress):
(WebKit::KeyAutoRepeatHandler::keyRelease):
* Source/WebKit/UIProcess/API/glib/KeyAutoRepeatHandler.h: Added.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseKeyPressEvent):
(webkitWebViewBaseKeyReleaseEvent):
(webkitWebViewBaseKeyPressed):
(webkitWebViewBaseKeyReleased):
(webkitWebViewBaseSynthesizeKeyEvent):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::handleKeyboardEvent):
(WKWPE::View::synthesizeCompositionKeyPress):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:

Canonical link: <a href="https://commits.webkit.org/261363@main">https://commits.webkit.org/261363@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11477343295634591942d5889741f54b438a78da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43848 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2734 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120112 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11545 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117069 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16202 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/31033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44790 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12946 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32374 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9364 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18903 "Built successfully") | [💥 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51959 "An unexpected error occured. Recent messages:OS: Monterey (12.6.3), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15419 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4312 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
OK, though I bet this doesn't need to be GLib-specific. Could probably put it in a cross-platform location.

carlosgcampos committed:
[GTK][WPE] Implement KeyboardEvent.repeat
https://bugs.webkit.org/show_bug.cgi?id=253507

Reviewed by Michael Catanzaro.

It's always false for GTK and WPE. This is also used since 259146@main
to decide when to use the keyboard scrolling animation for page scrolls.
Neither GTK nor libpwe provide information about whether a key press
event is auto-repeated or not, so this adds a simple helper class to
handle it.

* Source/WebKit/Shared/NativeWebKeyboardEvent.h:
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebKeyboardEvent.cpp:
(WebKit::WebKeyboardEvent::WebKeyboardEvent):
* Source/WebKit/Shared/WebKeyboardEvent.h:
* Source/WebKit/Shared/gtk/NativeWebKeyboardEventGtk.cpp:
(WebKit::NativeWebKeyboardEvent::NativeWebKeyboardEvent):
* Source/WebKit/Shared/gtk/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebKeyboardEvent):
* Source/WebKit/Shared/gtk/WebEventFactory.h:
* Source/WebKit/Shared/libwpe/NativeWebKeyboardEventLibWPE.cpp:
(WebKit::NativeWebKeyboardEvent::NativeWebKeyboardEvent):
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebKeyboardEvent):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/KeyAutoRepeatHandler.cpp: Added.
(WebKit::KeyAutoRepeatHandler::keyPress):
(WebKit::KeyAutoRepeatHandler::keyRelease):
* Source/WebKit/UIProcess/API/glib/KeyAutoRepeatHandler.h: Added.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseKeyPressEvent):
(webkitWebViewBaseKeyReleaseEvent):
(webkitWebViewBaseKeyPressed):
(webkitWebViewBaseKeyReleased):
(webkitWebViewBaseSynthesizeKeyEvent):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::handleKeyboardEvent):
(WKWPE::View::synthesizeCompositionKeyPress):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:

Canonical link: https://commits.webkit.org/261363@main
